runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -12643,7 +12659,13 @@ J9::X86::TreeEvaluator::generateTestAndReportFieldWatchInstructions(TR::CodeGene
    deps->stopAddingConditions();
    generateLabelInstruction(LABEL, node, endLabel, deps, cg);
 
-   if (isInstanceField || (!isResolved && isWrite) || isAOTCompile)
+   if (isInstanceField 
+      || (!isResolved && isWrite) 
+      || isAOTCompile
+#if defined(JITSERVER_SUPPORT)
+      || cg->comp()->isOutOfProcessCompilation()
```

Comment:
This check is used in the exact same way as `isAOTCompile` in the 3 locations where `isAOTCompile` is used inside this `generateTestAndReportFieldWatchInstructions`. Can we combine these two checks into a single boolean? And maybe we can rename the variable too. For example, we can replace:
```
bool isAOTCompile = cg->comp()->compileRelocatableCode();
```
with
```
bool fieldClassNeedsRelocation = cg->needClassAndMethodPointerRelocations()
```

I believe we should be able to use this `cg->needClassAndMethodPointerRelocations()` in the p and z codgens changes here as well, since the checks are all concerned with the `fieldClass` whose relo type is `TR_ClassAddress`

Reply:
is `isAOTCompile || cg->comp()->isOutOfProcessCompilation()` equivalent to `fieldClassNeesRelocation`? Wouldn't the first one evaluate to `true` and second one to `false` in this case?

Reply:
That's correct. The answer will be different. Here's my explanation for why we should be using `fieldClassNeedsRelocation`. If you spot anything wrong with it then please let me know. Otherwise if you can run a quick test to verify that it's working as expected then that'd be great!

The `isAOTCompile` boolean is used to indirectly check if `fieldClass` inside the snippet will be relocated. If it _will_ be relocated, then we can't use the address at compile time to generate optimal code. Instead, we must generate instructions to load the `fieldClass` (which would have been relocated correctly) into a register at runtime, and _then_ create a memory reference. This is why we have the code layout as follows in this routine:

```
if (!isAOTCompile)
   // use fieldClass at compile time to create a memory reference quickly --> this is the fast path
else
   // first load fieldClass from snippet into a register, then create the memory reference
```

So now the question is: When can we use the fast path? The answer is that we can use the fast path as long as we are not doing an AOT compile. This is because we know that the `fieldClass` will be correct for regular JIT compiles (obviously) and regular JITServer compiles (as we checked yesterday, `fieldClass` is obtained from `getDeclaringClassFromFieldOrStatic`, which is a remote query).

Last I checked, the `needClassAndMethodPointerRelocations` returns true for AOT compiles and AOTaaS compiles. Meanwhile, it returns false for JIT Compiles and remote JIT Compiles. So this query should provide us with the behaviour we want.

Reply:
I have verified that the new change is working as expected.